In [2]:
import pandas as pd
from sqlalchemy import create_engine

In [3]:
#importing the csv file contain the Winery's Lat and Long

csv_file = "Resources_weather/Location_df.csv"
Location_data_df = pd.read_csv(csv_file)
Location_data_df

Winery   Latitude   Longitude
0          Plantagenet -34.623193  117.662067
1               Xanadu -33.982649  115.056427
2          Vasse River -33.677252  119.188485
3        Robert Oatley -27.672817  121.628310
4              Picardy -34.395843  116.128722
5     Frankland Estate -34.462871  117.013378
6          Hope Estate -32.767729  151.312050
7       West Cape Howe -34.655733  117.557831
8       Leeuwin Estate -34.013389  115.065652
9     Wildberry Estate -27.672817  121.628310
10         Vasse Felix -33.822468  115.037726
11           Watershed -33.992480  115.090956
12           Clairault -33.732071  115.053474
13       Strait Jacket -27.672817  121.628310
14              Cullen -33.818829  115.037933
15      Voyager Estate -33.996743  115.054167
16             Element -27.672817  121.628310
17         Hamelin Bay -34.173626  115.121200
18        Stella Bella -33.972472  115.103043
19         Howard Park -33.831100  115.058846
20           Ironberry -27.672817  121.628310
21       Cape Mentelle -33.967055  115.037300
22           Streicker -33.732071  115.053474
23          Cloudburst -27.672817  121.628310
24         Rocky Gully -34.508684  117.008797
25             Juniper -33.824460  115.048465
26            Mad Fish -33.831211  115.058523
27         Skuttlebutt -27.672817  121.628310
28             Alkoomi -34.355877  116.964984
29        Devil's Lair -34.072623  115.107385
30          Beach Head -27.672817  121.628310
31           Moss Wood -33.783882  115.035811
32                Wise -33.579833  115.072491
33  Miles From Nowhere -27.672817  121.628310
34           Cherubino -33.745155  115.030126
35       Celestial Bay -32.015516  115.868497
36            Salitage -34.405875  116.128935
37   Snake and Herring -33.769747  115.035323
38      Chalice Bridge -27.672817  121.628310
39            Ringbolt -27.672817  121.628310
40      Evans and Tate -33.788298  115.034952
41              Ad Lib -27.672817  121.628310
42             Edwards -33.897537  115.030207
43        Fonty's Pool -34.289633  116.066141
44       Franklin Tate -27.672817  121.628310
45          Sandalford -31.868718  115.992708
46         Kiss Chasey -27.672817  121.628310
47              Merops -33.947899  115.069463
48            Killerby -33.813320  115.035953
49           Vinaceous -27.672817  121.628310
50             Marquee -27.672817  121.628310
51     Larry Cherubino -27.672817  121.628310
52          Capel Vale -33.541308  115.549871
53      Howling Wolves -31.789004  115.831612
54           Bush Bike -27.672817  121.628310
55               Happs -33.664926  115.095621
56      Xavier Flouret -27.672817  121.628310
57           Ferngrove -34.379623  116.933615
58    The Wishing Tree -27.672817  121.628310
59  Marchand and Burch -33.831211  115.058523

In [4]:
#round down those Lat and lng statistic to only 2 digits to look up weather stations from BOM
#http://www.bom.gov.au/climate/data/index.shtml?bookmark=136

In [5]:
Location_data_df.round({'Latitude':2,'Longitude':2})


Winery  Latitude  Longitude
0          Plantagenet    -34.62     117.66
1               Xanadu    -33.98     115.06
2          Vasse River    -33.68     119.19
3        Robert Oatley    -27.67     121.63
4              Picardy    -34.40     116.13
5     Frankland Estate    -34.46     117.01
6          Hope Estate    -32.77     151.31
7       West Cape Howe    -34.66     117.56
8       Leeuwin Estate    -34.01     115.07
9     Wildberry Estate    -27.67     121.63
10         Vasse Felix    -33.82     115.04
11           Watershed    -33.99     115.09
12           Clairault    -33.73     115.05
13       Strait Jacket    -27.67     121.63
14              Cullen    -33.82     115.04
15      Voyager Estate    -34.00     115.05
16             Element    -27.67     121.63
17         Hamelin Bay    -34.17     115.12
18        Stella Bella    -33.97     115.10
19         Howard Park    -33.83     115.06
20           Ironberry    -27.67     121.63
21       Cape Mentelle    -33.97     115.04
22           Streicker    -33.73     115.05
23          Cloudburst    -27.67     121.63
24         Rocky Gully    -34.51     117.01
25             Juniper    -33.82     115.05
26            Mad Fish    -33.83     115.06
27         Skuttlebutt    -27.67     121.63
28             Alkoomi    -34.36     116.96
29        Devil's Lair    -34.07     115.11
30          Beach Head    -27.67     121.63
31           Moss Wood    -33.78     115.04
32                Wise    -33.58     115.07
33  Miles From Nowhere    -27.67     121.63
34           Cherubino    -33.75     115.03
35       Celestial Bay    -32.02     115.87
36            Salitage    -34.41     116.13
37   Snake and Herring    -33.77     115.04
38      Chalice Bridge    -27.67     121.63
39            Ringbolt    -27.67     121.63
40      Evans and Tate    -33.79     115.03
41              Ad Lib    -27.67     121.63
42             Edwards    -33.90     115.03
43        Fonty's Pool    -34.29     116.07
44       Franklin Tate    -27.67     121.63
45          Sandalford    -31.87     115.99
46         Kiss Chasey    -27.67     121.63
47              Merops    -33.95     115.07
48            Killerby    -33.81     115.04
49           Vinaceous    -27.67     121.63
50             Marquee    -27.67     121.63
51     Larry Cherubino    -27.67     121.63
52          Capel Vale    -33.54     115.55
53      Howling Wolves    -31.79     115.83
54           Bush Bike    -27.67     121.63
55               Happs    -33.66     115.10
56      Xavier Flouret    -27.67     121.63
57           Ferngrove    -34.38     116.93
58    The Wishing Tree    -27.67     121.63
59  Marchand and Burch    -33.83     115.06

Using the above Lat and Lng details, we are able to find out all the weather station closest to the Location data within 50km 

#Using the Lat and Lng details, we noticed that Hope Estate is actually locate in NSW not WA. 

Trying to do the clean on one of the rainfall csv file, if it looks fine then apply the change to consolidated master file

In [6]:
#importing the weather station 9025 rainfall
csv_file_9025_rainfall = "Resources_weather/9025_rainfall.csv"
rainfall_9025_data_df = pd.read_csv(csv_file_9025_rainfall)
rainfall_9025_data_df

Product code  Station Number  Year   Jan   Feb   Mar   Apr    May    Jun  \
0     IDCJAC0001            9025  1886   0.0  10.4   0.0  11.4   66.0   88.9   
1     IDCJAC0001            9025  1887   2.5  13.0  21.8  59.4   77.5  138.9   
2     IDCJAC0001            9025  1914   NaN   NaN   NaN   NaN    NaN  112.1   
3     IDCJAC0001            9025  1915  75.3  82.8  36.0  32.1  126.5  203.5   
4     IDCJAC0001            9025  1916  34.9  43.5   9.9   1.8   96.1  208.1   
..           ...             ...   ...   ...   ...   ...   ...    ...    ...   
102   IDCJAC0001            9025  2017  54.7  66.8  26.0   0.0   73.3   42.4   
103   IDCJAC0001            9025  2018  76.2   0.0   0.0  23.2   66.5  142.4   
104   IDCJAC0001            9025  2019   1.0   0.0   1.6   0.5   39.2  190.6   
105   IDCJAC0001            9025  2020   0.0   0.0   NaN   NaN    NaN    NaN   
106   IDCJAC0001            9025  2021   0.0   0.0   2.2   0.0   48.0    NaN   

       Jul    Aug    Sep   Oct   Nov   Dec  Annual  
0    149.6  174.0  143.5  15.0  23.1   0.0   681.9  
1    221.5  147.1   88.1  37.6  32.8   9.1   849.3  
2    117.3   53.2   23.5  28.7  31.8  33.8     NaN  
3    237.7  166.5  130.6  83.6   7.6   1.8  1184.0  
4    175.2  175.2    2.8  55.6  69.4   0.0   872.5  
..     ...    ...    ...   ...   ...   ...     ...  
102  184.6  164.6   78.7  31.8   9.0  32.8   764.7  
103  144.9    NaN   28.2  36.6   1.0   1.4     NaN  
104   96.0   48.2   33.3  12.0  20.0   0.0   442.4  
105    NaN    NaN    NaN   NaN   NaN  45.0     NaN  
106    NaN    NaN    NaN   NaN   NaN   NaN     NaN  

[107 rows x 16 columns]

In [7]:
#Fill all the NA value with Zero
rainfall_9025_fillna_df = rainfall_9025_data_df.fillna(00)
rainfall_9025_fillna_df

Product code  Station Number  Year   Jan   Feb   Mar   Apr    May    Jun  \
0     IDCJAC0001            9025  1886   0.0  10.4   0.0  11.4   66.0   88.9   
1     IDCJAC0001            9025  1887   2.5  13.0  21.8  59.4   77.5  138.9   
2     IDCJAC0001            9025  1914   0.0   0.0   0.0   0.0    0.0  112.1   
3     IDCJAC0001            9025  1915  75.3  82.8  36.0  32.1  126.5  203.5   
4     IDCJAC0001            9025  1916  34.9  43.5   9.9   1.8   96.1  208.1   
..           ...             ...   ...   ...   ...   ...   ...    ...    ...   
102   IDCJAC0001            9025  2017  54.7  66.8  26.0   0.0   73.3   42.4   
103   IDCJAC0001            9025  2018  76.2   0.0   0.0  23.2   66.5  142.4   
104   IDCJAC0001            9025  2019   1.0   0.0   1.6   0.5   39.2  190.6   
105   IDCJAC0001            9025  2020   0.0   0.0   0.0   0.0    0.0    0.0   
106   IDCJAC0001            9025  2021   0.0   0.0   2.2   0.0   48.0    0.0   

       Jul    Aug    Sep   Oct   Nov   Dec  Annual  
0    149.6  174.0  143.5  15.0  23.1   0.0   681.9  
1    221.5  147.1   88.1  37.6  32.8   9.1   849.3  
2    117.3   53.2   23.5  28.7  31.8  33.8     0.0  
3    237.7  166.5  130.6  83.6   7.6   1.8  1184.0  
4    175.2  175.2    2.8  55.6  69.4   0.0   872.5  
..     ...    ...    ...   ...   ...   ...     ...  
102  184.6  164.6   78.7  31.8   9.0  32.8   764.7  
103  144.9    0.0   28.2  36.6   1.0   1.4     0.0  
104   96.0   48.2   33.3  12.0  20.0   0.0   442.4  
105    0.0    0.0    0.0   0.0   0.0  45.0     0.0  
106    0.0    0.0    0.0   0.0   0.0   0.0     0.0  

[107 rows x 16 columns]

In [8]:
#Correct the sumation for the Annual column after filling all NA value with Zero
rainfall_9025_fillna_df['Annual']=rainfall_9025_fillna_df[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec']].sum(axis=1)


In [9]:
rainfall_9025_fillna_df

Product code  Station Number  Year   Jan   Feb   Mar   Apr    May    Jun  \
0     IDCJAC0001            9025  1886   0.0  10.4   0.0  11.4   66.0   88.9   
1     IDCJAC0001            9025  1887   2.5  13.0  21.8  59.4   77.5  138.9   
2     IDCJAC0001            9025  1914   0.0   0.0   0.0   0.0    0.0  112.1   
3     IDCJAC0001            9025  1915  75.3  82.8  36.0  32.1  126.5  203.5   
4     IDCJAC0001            9025  1916  34.9  43.5   9.9   1.8   96.1  208.1   
..           ...             ...   ...   ...   ...   ...   ...    ...    ...   
102   IDCJAC0001            9025  2017  54.7  66.8  26.0   0.0   73.3   42.4   
103   IDCJAC0001            9025  2018  76.2   0.0   0.0  23.2   66.5  142.4   
104   IDCJAC0001            9025  2019   1.0   0.0   1.6   0.5   39.2  190.6   
105   IDCJAC0001            9025  2020   0.0   0.0   0.0   0.0    0.0    0.0   
106   IDCJAC0001            9025  2021   0.0   0.0   2.2   0.0   48.0    0.0   

       Jul    Aug    Sep   Oct   Nov   Dec  Annual  
0    149.6  174.0  143.5  15.0  23.1   0.0   681.9  
1    221.5  147.1   88.1  37.6  32.8   9.1   849.3  
2    117.3   53.2   23.5  28.7  31.8  33.8   400.4  
3    237.7  166.5  130.6  83.6   7.6   1.8  1184.0  
4    175.2  175.2    2.8  55.6  69.4   0.0   872.5  
..     ...    ...    ...   ...   ...   ...     ...  
102  184.6  164.6   78.7  31.8   9.0  32.8   764.7  
103  144.9    0.0   28.2  36.6   1.0   1.4   520.4  
104   96.0   48.2   33.3  12.0  20.0   0.0   442.4  
105    0.0    0.0    0.0   0.0   0.0  45.0    45.0  
106    0.0    0.0    0.0   0.0   0.0   0.0    50.2  

[107 rows x 16 columns]

In [10]:
#Groupby the Station Number for the average rainfall since year 2000
rainfall_9025_agg=rainfall_9025_fillna_df.loc[rainfall_9025_fillna_df["Year"]>=2000,["Station Number","Annual"]]
rainfall_9025_agg.groupby("Station Number").mean()

Annual
Station Number            
9025            605.068182

Creating a consolidated master file to group all the rainfall csv file into one and then apply the data cleaning and transformation

In [11]:
#Import os to interact with the underlying operating system/ returns the absolute path of the working directory
import os
os.getcwd()

'c:\\Users\\use\\ETL-project'

In [12]:
#Looping the Resources_weather folder for each file name under rainfall.csv and join them together into a master file

dfs=[]

for file in os.listdir("Resources_weather"):
   

    if file.endswith("rainfall.csv"):
        dfs.append(pd.read_csv(f"Resources_weather/{file}"))
        print(file)





10905_rainfall.csv
12280_rainfall.csv
12304_rainfall.csv
9025_rainfall.csv
9105_rainfall.csv
9172_rainfall.csv
9519_rainfall.csv
9547_rainfall.csv
9573_rainfall.csv
9574_rainfall.csv
9581_rainfall.csv
9592_rainfall.csv
9636_rainfall.csv
9661_rainfall.csv
9801_rainfall.csv
9843_rainfall.csv
9979_rainfall.csv
9992_rainfall.csv
rainfall.csv


In [13]:
rainfall_df=pd.concat(dfs)
rainfall_df.head()

Product code  Station Number    Year   Jan   Feb   Mar   Apr   May   Jun  \
0   IDCJAC0001         10905.0  1994.0   NaN   NaN   NaN   NaN   NaN   2.2   
1   IDCJAC0001         10905.0  1995.0  31.6  30.0  18.8   4.0  41.8  37.6   
2   IDCJAC0001         10905.0  1996.0   1.2  11.4   6.0  13.6  27.8   5.2   
3   IDCJAC0001         10905.0  1997.0   2.4  41.2  74.0  55.4  33.3  38.7   
4   IDCJAC0001         10905.0  1998.0   1.4   2.0  29.4  77.8  26.6  73.6   

    Jul   Aug   Sep   Oct   Nov   Dec  Annual  Closest_weather_station  \
0  18.8  23.4  26.6  48.6  15.6   9.8     NaN                      NaN   
1  60.0  22.0  56.2  46.6   9.0  67.0   424.6                      NaN   
2  93.4  43.8  73.6  26.4  27.0   6.2   335.6                      NaN   
3  50.0  59.6  65.9  24.1  64.9  14.0   523.5                      NaN   
4  37.8  77.7  29.0  49.0  18.8  10.4   433.5                      NaN   

   Average_yearly_rainfall  
0                      NaN  
1                      NaN  
2                      NaN  
3                      NaN  
4                      NaN

In [14]:
#Fill up all the NA value with Zero value and correct the summation of "Annual" column 

rainfall_clean_df = rainfall_df.fillna(00)

rainfall_clean_df['Annual']=rainfall_clean_df[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec']].sum(axis=1)
rainfall_clean_df

Product code  Station Number    Year   Jan   Feb   Mar   Apr   May   Jun  \
0    IDCJAC0001         10905.0  1994.0   0.0   0.0   0.0   0.0   0.0   2.2   
1    IDCJAC0001         10905.0  1995.0  31.6  30.0  18.8   4.0  41.8  37.6   
2    IDCJAC0001         10905.0  1996.0   1.2  11.4   6.0  13.6  27.8   5.2   
3    IDCJAC0001         10905.0  1997.0   2.4  41.2  74.0  55.4  33.3  38.7   
4    IDCJAC0001         10905.0  1998.0   1.4   2.0  29.4  77.8  26.6  73.6   
..          ...             ...     ...   ...   ...   ...   ...   ...   ...   
12            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0   0.0   
13            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0   0.0   
14            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0   0.0   
15            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0   0.0   
16            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0   0.0   

     Jul   Aug   Sep   Oct   Nov   Dec  Annual  Closest_weather_station  \
0   18.8  23.4  26.6  48.6  15.6   9.8   145.0                      0.0   
1   60.0  22.0  56.2  46.6   9.0  67.0   424.6                      0.0   
2   93.4  43.8  73.6  26.4  27.0   6.2   335.6                      0.0   
3   50.0  59.6  65.9  24.1  64.9  14.0   523.5                      0.0   
4   37.8  77.7  29.0  49.0  18.8  10.4   433.5                      0.0   
..   ...   ...   ...   ...   ...   ...     ...                      ...   
12   0.0   0.0   0.0   0.0   0.0   0.0     0.0                   9843.0   
13   0.0   0.0   0.0   0.0   0.0   0.0     0.0                   9978.0   
14   0.0   0.0   0.0   0.0   0.0   0.0     0.0                   9992.0   
15   0.0   0.0   0.0   0.0   0.0   0.0     0.0                  10905.0   
16   0.0   0.0   0.0   0.0   0.0   0.0     0.0                  12304.0   

    Average_yearly_rainfall  
0                      0.00  
1                      0.00  
2                      0.00  
3                      0.00  
4                      0.00  
..                      ...  
12                   524.84  
13                   722.28  
14                   645.66  
15                   428.32  
16                   236.58  

[1359 rows x 18 columns]

In [15]:
#Selecting only year older than 2000, and column of station number and annual
rainfall_data_df=rainfall_clean_df.loc[rainfall_clean_df["Year"]>=2000,["Station Number","Annual"]]
rainfall_data_df

Station Number  Annual
6          10905.0   413.2
7          10905.0   505.6
8          10905.0   267.0
9          10905.0   618.4
10         10905.0   397.6
..             ...     ...
15          9992.0   619.6
16          9992.0   539.8
17          9992.0   722.2
18          9992.0   765.8
19          9992.0   106.4

[405 rows x 2 columns]

The weather station 12280 has faulty data, so will use the substitute weather station 12304 which we downloaded as a backup previously. 
Weather station 12304 is 2nd closest station for the Lat / Lng location

In [16]:
#Drop off weather station 12280 as data are faulty
rainfall_data_excl_df = rainfall_data_df[rainfall_data_df["Station Number"] != 12280]
rainfall_data_excl_df

Station Number  Annual
6          10905.0   413.2
7          10905.0   505.6
8          10905.0   267.0
9          10905.0   618.4
10         10905.0   397.6
..             ...     ...
15          9992.0   619.6
16          9992.0   539.8
17          9992.0   722.2
18          9992.0   765.8
19          9992.0   106.4

[382 rows x 2 columns]

In [17]:
#check how many station left
No_of_station = len(rainfall_data_excl_df["Station Number"].unique())
No_of_station

17

In [18]:
#Name change for "Annual Column" to "Annual Average Rainfall (mm)"
rainfall_data_namechange_df=rainfall_data_excl_df.rename(columns={"Annual":"Average_yearly_rainfall","Station Number":"Closest_weather_station"})
rainfall_data_namechange_df

Closest_weather_station  Average_yearly_rainfall
6                   10905.0                    413.2
7                   10905.0                    505.6
8                   10905.0                    267.0
9                   10905.0                    618.4
10                  10905.0                    397.6
..                      ...                      ...
15                   9992.0                    619.6
16                   9992.0                    539.8
17                   9992.0                    722.2
18                   9992.0                    765.8
19                   9992.0                    106.4

[382 rows x 2 columns]

In [19]:
#Groupby the station number and calculate the mean for year older than 2000
rainfall_data_average_df=rainfall_data_namechange_df.groupby("Closest_weather_station").mean()
rainfall_data_average_df

Average_yearly_rainfall
Closest_weather_station                         
9025.0                                605.068182
9105.0                                624.326087
9172.0                                720.443478
9519.0                                643.965217
9547.0                                999.091304
9573.0                                797.552174
9574.0                                955.995652
9581.0                                596.843478
9592.0                               1071.504348
9636.0                                965.769565
9661.0                                639.509091
9801.0                                840.926087
9843.0                                524.843478
9978.0                                722.276190
9992.0                                645.660000
10905.0                               428.321739
12304.0                               236.576190

In [20]:
#Round to 2 decimal place for better presentation
rainfall_data_average_FINAL_df=rainfall_data_average_df.round({'Average_yearly_rainfall':2})
rainfall_data_average_FINAL_df

Average_yearly_rainfall
Closest_weather_station                         
9025.0                                    605.07
9105.0                                    624.33
9172.0                                    720.44
9519.0                                    643.97
9547.0                                    999.09
9573.0                                    797.55
9574.0                                    956.00
9581.0                                    596.84
9592.0                                   1071.50
9636.0                                    965.77
9661.0                                    639.51
9801.0                                    840.93
9843.0                                    524.84
9978.0                                    722.28
9992.0                                    645.66
10905.0                                   428.32
12304.0                                   236.58

In [51]:
rainfall_FINAL_DF=rainfall_data_average_FINAL_df.reset_index()
rainfall_FINAL_DF

Closest_weather_station  Average_yearly_rainfall
0                    9025.0                   605.07
1                    9105.0                   624.33
2                    9172.0                   720.44
3                    9519.0                   643.97
4                    9547.0                   999.09
5                    9573.0                   797.55
6                    9574.0                   956.00
7                    9581.0                   596.84
8                    9592.0                  1071.50
9                    9636.0                   965.77
10                   9661.0                   639.51
11                   9801.0                   840.93
12                   9843.0                   524.84
13                   9978.0                   722.28
14                   9992.0                   645.66
15                  10905.0                   428.32
16                  12304.0                   236.58

In [53]:
#Export to csv file
rainfall_FINAL_DF.to_csv("Resources/weather/rainfall.csv", index=True)

In [22]:
#Looping the Resources_weather folder for each file name under solar.csv and join them together into a master file

dfs=[]

for file in os.listdir("Resources_weather"):
   

    if file.endswith("solar.csv"):
        dfs.append(pd.read_csv(f"Resources_weather/{file}"))
        print(file)



10905_solar.csv
12280_solar.csv
12304_solar.csv
9025_solar.csv
9105_solar.csv
9172_solar.csv
9519_solar.csv
9547_solar.csv
9573_solar.csv
9574_solar.csv
9581_solar.csv
9592_solar.csv
9636_solar.csv
9661_solar.csv
9801_solar.csv
9843_solar.csv
9978_solar.csv
9992_solar.csv
solar.csv


In [23]:
solar_df=pd.concat(dfs)
solar_df

Product code  Station Number    Year   Jan   Feb   Mar   Apr   May  Jun  \
0    IDCJAC0003         10905.0  1990.0  25.7  22.6  14.7  12.7   9.6  8.5   
1    IDCJAC0003         10905.0  1991.0  24.8  22.3  18.0  13.5  10.2  8.1   
2    IDCJAC0003         10905.0  1992.0  26.4  20.1  14.4  12.5  10.1  8.2   
3    IDCJAC0003         10905.0  1993.0  29.0  22.2  16.2  13.5   9.0  8.2   
4    IDCJAC0003         10905.0  1994.0  28.4  22.3  18.4  14.7  10.1  9.0   
..          ...             ...     ...   ...   ...   ...   ...   ...  ...   
12          NaN             NaN     NaN   NaN   NaN   NaN   NaN   NaN  NaN   
13          NaN             NaN     NaN   NaN   NaN   NaN   NaN   NaN  NaN   
14          NaN             NaN     NaN   NaN   NaN   NaN   NaN   NaN  NaN   
15          NaN             NaN     NaN   NaN   NaN   NaN   NaN   NaN  NaN   
16          NaN             NaN     NaN   NaN   NaN   NaN   NaN   NaN  NaN   

    Jul   Aug   Sep   Oct   Nov   Dec  Annual  Closest_weather_station  \
0   9.5  11.7  16.4  18.0  27.6  24.5    16.8                      NaN   
1   8.6  12.2  15.8  21.3  24.1  24.6    17.0                      NaN   
2   8.9  10.7  13.0  19.6  19.8  27.6    15.9                      NaN   
3   8.9  10.8  16.3  20.3  21.5  28.9    17.1                      NaN   
4   9.7  14.6  16.3  18.6  23.6  28.7    17.9                      NaN   
..  ...   ...   ...   ...   ...   ...     ...                      ...   
12  NaN   NaN   NaN   NaN   NaN   NaN     NaN                   9843.0   
13  NaN   NaN   NaN   NaN   NaN   NaN     NaN                   9978.0   
14  NaN   NaN   NaN   NaN   NaN   NaN     NaN                   9992.0   
15  NaN   NaN   NaN   NaN   NaN   NaN     NaN                  10905.0   
16  NaN   NaN   NaN   NaN   NaN   NaN     NaN                  12304.0   

    Average_yearly_solar_exposure  
0                             NaN  
1                             NaN  
2                             NaN  
3                             NaN  
4                             NaN  
..                            ...  
12                         194.64  
13                         207.31  
14                         212.70  
15                         200.85  
16                         240.68  

[611 rows x 18 columns]

In [24]:
#Fill up all the NA value with Zero value and correct the summation of "Annual" column 


solar_clean_df = solar_df.fillna(00)

solar_clean_df['Annual']=solar_clean_df[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec']].sum(axis=1)
solar_clean_df

Product code  Station Number    Year   Jan   Feb   Mar   Apr   May  Jun  \
0    IDCJAC0003         10905.0  1990.0  25.7  22.6  14.7  12.7   9.6  8.5   
1    IDCJAC0003         10905.0  1991.0  24.8  22.3  18.0  13.5  10.2  8.1   
2    IDCJAC0003         10905.0  1992.0  26.4  20.1  14.4  12.5  10.1  8.2   
3    IDCJAC0003         10905.0  1993.0  29.0  22.2  16.2  13.5   9.0  8.2   
4    IDCJAC0003         10905.0  1994.0  28.4  22.3  18.4  14.7  10.1  9.0   
..          ...             ...     ...   ...   ...   ...   ...   ...  ...   
12            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0  0.0   
13            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0  0.0   
14            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0  0.0   
15            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0  0.0   
16            0             0.0     0.0   0.0   0.0   0.0   0.0   0.0  0.0   

    Jul   Aug   Sep   Oct   Nov   Dec  Annual  Closest_weather_station  \
0   9.5  11.7  16.4  18.0  27.6  24.5   201.5                      0.0   
1   8.6  12.2  15.8  21.3  24.1  24.6   203.5                      0.0   
2   8.9  10.7  13.0  19.6  19.8  27.6   191.3                      0.0   
3   8.9  10.8  16.3  20.3  21.5  28.9   204.8                      0.0   
4   9.7  14.6  16.3  18.6  23.6  28.7   214.4                      0.0   
..  ...   ...   ...   ...   ...   ...     ...                      ...   
12  0.0   0.0   0.0   0.0   0.0   0.0     0.0                   9843.0   
13  0.0   0.0   0.0   0.0   0.0   0.0     0.0                   9978.0   
14  0.0   0.0   0.0   0.0   0.0   0.0     0.0                   9992.0   
15  0.0   0.0   0.0   0.0   0.0   0.0     0.0                  10905.0   
16  0.0   0.0   0.0   0.0   0.0   0.0     0.0                  12304.0   

    Average_yearly_solar_exposure  
0                            0.00  
1                            0.00  
2                            0.00  
3                            0.00  
4                            0.00  
..                            ...  
12                         194.64  
13                         207.31  
14                         212.70  
15                         200.85  
16                         240.68  

[611 rows x 18 columns]

In [25]:
#Selecting only year older than 2000, and column of station number and annual
solar_data_df=solar_clean_df.loc[solar_clean_df["Year"]>=2000,["Station Number","Annual"]]
solar_data_df

Station Number  Annual
10         10905.0   214.3
11         10905.0   207.2
12         10905.0   212.1
13         10905.0   207.0
14         10905.0   208.5
..             ...     ...
28          9992.0   212.7
29          9992.0   222.4
30          9992.0   214.7
31          9992.0   211.6
32          9992.0   105.5

[414 rows x 2 columns]

In [26]:
#Drop off weather station 12280 as data are faulty
solar_data_excl_df = solar_data_df[solar_data_df["Station Number"] != 12280]
solar_data_excl_df

Station Number  Annual
10         10905.0   214.3
11         10905.0   207.2
12         10905.0   212.1
13         10905.0   207.0
14         10905.0   208.5
..             ...     ...
28          9992.0   212.7
29          9992.0   222.4
30          9992.0   214.7
31          9992.0   211.6
32          9992.0   105.5

[391 rows x 2 columns]

In [27]:
#check how many station left
No_of_station = len(solar_data_excl_df["Station Number"].unique())
No_of_station

17

In [28]:
#Name change for "Annual Column" to "Annual Average Solar Exposure (MJ m-2)"
solar_data_namechange_df=solar_data_excl_df.rename(columns={"Annual":"Average_yearly_solar_exposure","Station Number":"Closest_weather_station"})
solar_data_namechange_df

Closest_weather_station  Average_yearly_solar_exposure
10                  10905.0                          214.3
11                  10905.0                          207.2
12                  10905.0                          212.1
13                  10905.0                          207.0
14                  10905.0                          208.5
..                      ...                            ...
28                   9992.0                          212.7
29                   9992.0                          222.4
30                   9992.0                          214.7
31                   9992.0                          211.6
32                   9992.0                          105.5

[391 rows x 2 columns]

In [29]:
#Groupby the station number and calculate the mean for years older than 2000
solar_data_average_df=solar_data_namechange_df.groupby("Closest_weather_station").mean()
solar_data_average_df

Average_yearly_solar_exposure
Closest_weather_station                               
9025.0                                      222.108696
9105.0                                      224.660870
9172.0                                      222.860870
9519.0                                      214.730435
9547.0                                      198.917391
9573.0                                      196.304348
9574.0                                      202.434783
9581.0                                      192.104348
9592.0                                      190.600000
9636.0                                      204.834783
9661.0                                      193.117391
9801.0                                      196.373913
9843.0                                      194.643478
9978.0                                      207.313043
9992.0                                      212.695652
10905.0                                     200.852174
12304.0                                     240.682609

In [30]:
#Round to 2 decimal place for better presentation
solar_data_average_FINAL_df=solar_data_average_df.round({'Average_yearly_solar_exposure':2})
solar_data_average_FINAL_df

Average_yearly_solar_exposure
Closest_weather_station                               
9025.0                                          222.11
9105.0                                          224.66
9172.0                                          222.86
9519.0                                          214.73
9547.0                                          198.92
9573.0                                          196.30
9574.0                                          202.43
9581.0                                          192.10
9592.0                                          190.60
9636.0                                          204.83
9661.0                                          193.12
9801.0                                          196.37
9843.0                                          194.64
9978.0                                          207.31
9992.0                                          212.70
10905.0                                         200.85
12304.0                                         240.68

In [50]:
Solar_FINAL_df=solar_data_average_FINAL_df.reset_index()
Solar_FINAL_df

Closest_weather_station  Average_yearly_solar_exposure
0                    9025.0                         222.11
1                    9105.0                         224.66
2                    9172.0                         222.86
3                    9519.0                         214.73
4                    9547.0                         198.92
5                    9573.0                         196.30
6                    9574.0                         202.43
7                    9581.0                         192.10
8                    9592.0                         190.60
9                    9636.0                         204.83
10                   9661.0                         193.12
11                   9801.0                         196.37
12                   9843.0                         194.64
13                   9978.0                         207.31
14                   9992.0                         212.70
15                  10905.0                         200.85
16                  12304.0                         240.68

In [54]:
#Export the result to csv file
Solar_FINAL_df.to_csv("Resources/weather/solar.csv", index=True)

In [32]:
csv_file = "Resources_weather/Location_with_weather_station.csv"
Location_weather_df = pd.read_csv(csv_file)
Location_weather_df



Unnamed: 0              Winery  Latitude  Longitude  \
0            0         Plantagenet    -34.62     117.66   
1            1              Xanadu    -33.98     115.06   
2            2         Vasse River    -33.68     119.19   
3            3       Robert Oatley    -27.67     121.63   
4            4             Picardy    -34.40     116.13   
5            5    Frankland Estate    -34.46     117.01   
6            6         Hope Estate    -32.77     151.31   
7            7      West Cape Howe    -34.66     117.56   
8            8      Leeuwin Estate    -34.01     115.07   
9            9    Wildberry Estate    -27.67     121.63   
10          10         Vasse Felix    -33.82     115.04   
11          11           Watershed    -33.99     115.09   
12          12           Clairault    -33.73     115.05   
13          13       Strait Jacket    -27.67     121.63   
14          14              Cullen    -33.82     115.04   
15          15      Voyager Estate    -34.00     115.05   
16          16             Element    -27.67     121.63   
17          17         Hamelin Bay    -34.17     115.12   
18          18        Stella Bella    -33.97     115.10   
19          19         Howard Park    -33.83     115.06   
20          20           Ironberry    -27.67     121.63   
21          21       Cape Mentelle    -33.97     115.04   
22          22           Streicker    -33.73     115.05   
23          23          Cloudburst    -27.67     121.63   
24          24         Rocky Gully    -34.51     117.01   
25          25             Juniper    -33.82     115.05   
26          26            Mad Fish    -33.83     115.06   
27          27         Skuttlebutt    -27.67     121.63   
28          28             Alkoomi    -34.36     116.96   
29          29        Devil's Lair    -34.07     115.11   
30          30          Beach Head    -27.67     121.63   
31          31           Moss Wood    -33.78     115.04   
32          32                Wise    -33.58     115.07   
33          33  Miles From Nowhere    -27.67     121.63   
34          34           Cherubino    -33.75     115.03   
35          35       Celestial Bay    -32.02     115.87   
36          36            Salitage    -34.41     116.13   
37          37   Snake and Herring    -33.77     115.04   
38          38      Chalice Bridge    -27.67     121.63   
39          39            Ringbolt    -27.67     121.63   
40          40      Evans and Tate    -33.79     115.03   
41          41              Ad Lib    -27.67     121.63   
42          42             Edwards    -33.90     115.03   
43          43        Fonty's Pool    -34.29     116.07   
44          44       Franklin Tate    -27.67     121.63   
45          45          Sandalford    -31.87     115.99   
46          46         Kiss Chasey    -27.67     121.63   
47          47              Merops    -33.95     115.07   
48          48            Killerby    -33.81     115.04   
49          49           Vinaceous    -27.67     121.63   
50          50             Marquee    -27.67     121.63   
51          51     Larry Cherubino    -27.67     121.63   
52          52          Capel Vale    -33.54     115.55   
53          53      Howling Wolves    -31.79     115.83   
54          54           Bush Bike    -27.67     121.63   
55          55               Happs    -33.66     115.10   
56          56      Xavier Flouret    -27.67     121.63   
57          57           Ferngrove    -34.38     116.93   
58          58    The Wishing Tree    -27.67     121.63   
59          59  Marchand and Burch    -33.83     115.06   

    Closest_weather_station  Name_of_weather_station  
0                    9581.0         Mount Barker, WA  
1                    9574.0       Margaret River, WA  
2                   10905.0                Jacup, WA  
3                   12304.0              Melrose, WA  
4                    9592.0            Pemberton, WA  
5                    9843.0  Frankland Vineyards, WA  
6                  

In [33]:
Location_weather_Final_df=Location_weather_df[["Winery","Latitude","Longitude","Closest_weather_station","Name_of_weather_station"]]
Location_weather_Final_df.head()

Winery  Latitude  Longitude  Closest_weather_station  \
0    Plantagenet    -34.62     117.66                   9581.0   
1         Xanadu    -33.98     115.06                   9574.0   
2    Vasse River    -33.68     119.19                  10905.0   
3  Robert Oatley    -27.67     121.63                  12304.0   
4        Picardy    -34.40     116.13                   9592.0   

  Name_of_weather_station  
0        Mount Barker, WA  
1      Margaret River, WA  
2               Jacup, WA  
3             Melrose, WA  
4           Pemberton, WA

In [35]:
Location_weather_F_df = Location_weather_Final_df.drop(6)
Location_weather_F_df

Winery  Latitude  Longitude  Closest_weather_station  \
0          Plantagenet    -34.62     117.66                   9581.0   
1               Xanadu    -33.98     115.06                   9574.0   
2          Vasse River    -33.68     119.19                  10905.0   
3        Robert Oatley    -27.67     121.63                  12304.0   
4              Picardy    -34.40     116.13                   9592.0   
5     Frankland Estate    -34.46     117.01                   9843.0   
7       West Cape Howe    -34.66     117.56                   9581.0   
8       Leeuwin Estate    -34.01     115.07                   9574.0   
9     Wildberry Estate    -27.67     121.63                  12304.0   
10         Vasse Felix    -33.82     115.04                   9636.0   
11           Watershed    -33.99     115.09                   9574.0   
12           Clairault    -33.73     115.05                   9636.0   
13       Strait Jacket    -27.67     121.63                  12304.0   
14              Cullen    -33.82     115.04                   9636.0   
15      Voyager Estate    -34.00     115.05                   9574.0   
16             Element    -27.67     121.63                  12304.0   
17         Hamelin Bay    -34.17     115.12                   9801.0   
18        Stella Bella    -33.97     115.10                   9574.0   
19         Howard Park    -33.83     115.06                   9636.0   
20           Ironberry    -27.67     121.63                  12304.0   
21       Cape Mentelle    -33.97     115.04                   9574.0   
22           Streicker    -33.73     115.05                   9636.0   
23          Cloudburst    -27.67     121.63                  12304.0   
24         Rocky Gully    -34.51     117.01                   9661.0   
25             Juniper    -33.82     115.05                   9636.0   
26            Mad Fish    -33.83     115.06                   9636.0   
27         Skuttlebutt    -27.67     121.63                  12304.0   
28             Alkoomi    -34.36     116.96                   9843.0   
29        Devil's Lair    -34.07     115.11                   9547.0   
30          Beach Head    -27.67     121.63                  12304.0   
31           Moss Wood    -33.78     115.04                   9636.0   
32                Wise    -33.58     115.07                   9519.0   
33  Miles From Nowhere    -27.67     121.63                  12304.0   
34           Cherubino    -33.75     115.03                   9636.0   
35       Celestial Bay    -32.02     115.87                   9172.0   
36            Salitage    -34.41     116.13                   9592.0   
37   Snake and Herring    -33.77     115.04                   9636.0   
38      Chalice Bridge    -27.67     121.63                  12304.0   
39            Ringbolt    -27.67     121.63                  12304.0   
40      Evans and Tate    -33.79     115.03                   9636.0   
41              Ad Lib    -27.67     121.63                  12304.0   
42             Edwards    -33.90     115.03                   9574.0   
43        Fonty's Pool    -34.29     116.07                   9573.0   
44       Franklin Tate    -27.67     121.63                  12304.0   
45          Sandalford    -31.87     115.99                   9025.0   
46         Kiss Chasey    -27.67     121.63                  12304.0   
47              Merops    -33.95     115.07                   9574.0   
48            Killerby    -33.81     115.04                   9636.0   
49           Vinaceous    -27.67     121.63                  12304.0   
50             Marquee    -27.67     121.63                  12304.0   
51     Larry Cherubino    -27.67     121.63                  12304.0   
52          Capel Vale    -33.54     115.55                   9992.0   
53      Howling Wolves    -31.79     115.83                   9105.0   
54           Bush Bike    -27.67     121.63                  12304.0   
55               Happs    -33.66     115.10             

In [39]:
user_name = "postgres"
password = "postgres"
rds_connection_string = f"{user_name}:{password}@localhost:5432/wa_wine_db"
engine = create_engine(f'postgresql://{rds_connection_string}')

In [42]:
# Check for tables
engine.table_names()

C:\Users\use\AppData\Local\Temp/ipykernel_18384/2986013912.py:2: SADeprecationWarning: The Engine.table_names() method is deprecated and will be removed in a future release.  Please refer to Inspector.get_table_names(). (deprecated since: 1.4)
  engine.table_names()


['Winery_Location',
 'Weather_Average_Rainfall',
 'Weather_Average_Solar',
 'Soil_data',
 'Winery_Export_Sales_by_Geolocation',
 'Wine_Volume_by_region',
 'Wine_Grape_variety_Sales',
 'Wine_Grape_variety_Production',
 'Winery_Tasting_Price_data']

In [55]:
# Use pandas to load csv converted dataframe
rainfall_FINAL_DF.to_sql(name='Weather_Average_Rainfall', con=engine, if_exists='append', index=False)

In [56]:
# Confirm data has been added by querying the table
pd.read_sql_query('select * from Weather_Average_Rainfall', con=engine).head()

ProgrammingError: (psycopg2.errors.UndefinedTable) relation "weather_average_rainfall" does not exist
LINE 1: select * from Weather_Average_Rainfall
                      ^

[SQL: select * from Weather_Average_Rainfall]
(Background on this error at: https://sqlalche.me/e/14/f405)

In [57]:
# Use pandas to load csv converted dataframe
Solar_FINAL_df.to_sql(name='Weather_Average_Solar', con=engine, if_exists='append', index=False)

In [58]:
pd.read_sql_query('select * from Weather_Average_Solar', con=engine).head()

ProgrammingError: (psycopg2.errors.UndefinedTable) relation "weather_average_solar" does not exist
LINE 1: select * from Weather_Average_Solar
                      ^

[SQL: select * from Weather_Average_Solar]
(Background on this error at: https://sqlalche.me/e/14/f405)